pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,406 @@
+		// The release of the identitiy allocated has been handled above. This
+		// can happen if we discover that the prefix is also associated with
+		// other labels besides kube-apiserver, i.e. if the kube-apiserver is
+		// not running outside of the cluster.
+		return injectLabelsForCIDR(prefix, lbls)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.IPAllocationTimeout)
+	defer cancel()
+	return IdentityAllocator.AllocateIdentity(ctx, lbls, false)
+}
+
+// injectLabelsForCIDR will allocate a CIDR identity for the given prefix. The
+// release of the identity must be managed by the caller.
+func injectLabelsForCIDR(prefix string, lbls labels.Labels) (*identity.Identity, bool, error) {
+	if !strings.Contains(prefix, "/") {
+		prefix = prefix + "/32"
+	}
+	_, cidr, err := net.ParseCIDR(prefix)
+	if err != nil {
+		return nil, false, err
+	}
+
+	allLbls := cidrlabels.GetCIDRLabels(cidr)
+	allLbls.MergeLabels(lbls)
+
+	log.WithFields(logrus.Fields{
+		logfields.CIDR:   cidr,
+		logfields.Labels: lbls, // omitting allLbls as CIDR labels would make this massive
+	}).Debug(
+		"Injecting CIDR labels for prefix",
+	)
+
+	return allocate(cidr, allLbls)
+}
+
+// FilterMetadataByLabels returns all the prefixes inside the IdentityMetadata
+// map which contain the given labels. Note that `filter` is a subset match,
+// not a full match.
+func FilterMetadataByLabels(filter labels.Labels) []string {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	matching := make([]string, 0)
+	for prefix, lbls := range IdentityMetadata {
+		if bytes.Contains(lbls.SortedList(), filter.SortedList()) {
+			matching = append(matching, prefix)
+		}
+	}
+	return matching
+}
+
+// RemoveAllPrefixesWithLabels wraps RemoveLabels to provide a convenient
+// method for the caller to remove all given prefixes at once. This function
+// will trigger policy update and recalculation if necessary on behalf of the
+// caller if any changes to the kube-apiserver were detected.
+//
+// Identities allocated by InjectLabels() may be released by RemoveLabels().
+//
+// A prefix will only be removed from the IDMD if the set of labels becomes
+// empty.
+func RemoveAllPrefixesWithLabels(
+	m map[string]labels.Labels,
+	src source.Source,
+	updater identityUpdater,
+	triggerer policyTriggerer,
+) {
+	var (
+		hasKubeAPIServer bool
+
+		idsToAdd    = make(map[identity.NumericIdentity]labels.LabelArray)
+		idsToDelete = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range m {
+		id, exists := IPIdentityCache.LookupByIP(prefix)
+		if id.ID <= identity.IdentityUnknown || !exists {
+			continue
+		}
+		if lbls.Has(labels.LabelKubeAPIServer[labels.IDNameKubeAPIServer]) {
+			hasKubeAPIServer = true
+		}
+		// Insert to propagate the updated set of labels after removal.
+		lbls := RemoveLabels(prefix, lbls, src).LabelArray()
+		idsToDelete[id.ID] = lbls
+		if len(lbls) > 0 {
+			idsToAdd[id.ID] = lbls
+		}
```

Comment:
This code is a little bit odd without comments it's hard to understand why one would like to add `lbls` to both `idsToDelete` and `idsToAdd`

Reply:
yeah it's a bit hard to parse. `RemoveLabels` will remove some subset of the labels. If the resulting set is empty, then that means there are no labels left to associate with the IP and therefore it should be removed. If the resulting set is non-empty, then we've basically removed some of the labels from being associated with the IP, but we have another subset of the labels that should still be associated with the IP. So we want to re-add those ones back in.

Consider if, for instance, kube-apiserver is running on one of the nodes in the cluster, then it gets redeployed. We'll execute a removal of the `kube-apiserver` label from this IP, but it will still be a `remote-node`. So then we remove the identity association that points the prefix -> identity(`remote-node`+`kube-apiserver`) and then insert  prefix -> identity(`remote-node`).

Reply:
That was my suspicion, can you please add a similar comment in the code?